File: Echowaves/EchoWaveViewController.m
//
//  WaveViewController.m
//  Echowaves
//
//  Created by Dmitry on 1/17/14.
//  Copyright (c) 2014 Echowaves. All rights reserved.
//

#import "EchoWaveViewController.h"
#import "NavigationTabBarViewController.h"
#import "EWImage.h"
#import "DetailedImagePageViewController.h"


@implementation EchoWaveViewController



- (void)viewDidLoad
{
    [super viewDidLoad];
    NSLog(@"$$$$$$$$$$$$$$$$calling viewDidLoad for EchoWaveViewController");

    
    [self emptyWaveLabel].hidden = YES;
    
    
    if(self.refreshControl == nil) {
        self.refreshControl = [UIRefreshControl new];
        [self.refreshControl addTarget:self action:@selector(startRefresh:)
                      forControlEvents:UIControlEventValueChanged];
        [self.imagesCollectionView addSubview:self.refreshControl];
    }
    
    [self startRefresh:self.refreshControl];
    
    [self reloadWavesPicker];


    

}

- (void) reloadWavesPicker {
    [EWWave getAllMyWaves:^(NSArray *waves) {
        self.myWaves = waves;
        
        NSLog(@"11111111111 currentWaveName: %@", [APP_DELEGATE currentWaveName]);
        
        if( [APP_DELEGATE currentWaveName] == NULL) {
            NSURLCredential *credential = [EWWave getStoredCredential];
            APP_DELEGATE.currentWaveName = [credential user];
            APP_DELEGATE.currentWaveIndex = 0;
        }
        NSLog(@"2222222222 currentWaveName: %@", [APP_DELEGATE currentWaveName]);
        
        
        NSLog(@"3333333333 wavesPickerSize: %lu", (unsigned long)self.myWaves.count);
        
        
        self.wavesPicker = [[UIPickerView alloc] initWithFrame:CGRectZero];
        [self attachPickerToTextField:self.waveSelected :self.wavesPicker];
        
//        [self.wavesPicker selectRow:APP_DELEGATE.currentWaveIndex inComponent:0 animated:NO];
        
        NSLog(@"setting wave index: %ld", [APP_DELEGATE currentWaveIndex]);
        self.navigationController.navigationBar.topItem.title = @"";//[APP_DELEGATE currentWaveName];
        //        [self.wavesPicker selectRow:[APP_DELEGATE currentWaveIndex] inComponent:0 animated:NO];
        
    } failure:^(NSError *error) {
        [EWWave showErrorAlertWithMessage:error.description
                               FromSender:nil];
    }];
}


- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    self.imagesCollectionView.autoresizingMask = UIViewAutoresizingFlexibleHeight |
    UIViewAutoresizingFlexibleWidth;
}

- (void) startRefresh:(UIRefreshControl *)sender {
    NSLog(@"starting the refresh");
    
    [EWImage getAllImagesForWave:[APP_DELEGATE currentWaveName]
                         success:^(NSArray *waveImages) {
                             self.waveImages = waveImages;
                             NSLog(@"@total images %lu", (unsigned long)[self.waveImages count]);
                             [self.imagesCollectionView reloadData];
                             [self.imagesCollectionView reloadInputViews];
                             if(waveImages.count == 0) {
                                 [self emptyWaveLabel].hidden = NO;
                             } else {
                                 [self emptyWaveLabel].hidden = YES;
                             }
                             [sender endRefreshing];
                         }
                         failure:^(NSError *error) {
                             NSLog(@"error %@", error.description);
                         }];
    
 
}

- (void)attachPickerToTextField: (UITextField*) textField :(UIPickerView*) picker{
    picker.delegate = self;
    picker.dataSource = self;
    
    textField.delegate = self;
    textField.inputView = picker;
}

#pragma mark - Picker delegate stuff

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView
{
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component
{
    NSLog(@"5555555555555 numberOfRowsInComponent: %lu", (unsigned long)self.myWaves.count);
    return self.myWaves.count;
}

#pragma mark -  UIPickerViewDelegate
-(UIView *)pickerView:(UIPickerView *)pickerView viewForRow:(NSInteger)row forComponent:(NSInteger)component reusingView:(UIView *)view
{
    UILabel *label = [[UILabel alloc] init];
    label.backgroundColor = [UIColor orangeColor];
    label.textColor = [UIColor whiteColor];
    label.font = [UIFont fontWithName:@"HelveticaNeue" size:14];
    label.textAlignment = NSTextAlignmentCenter; 
    //WithFrame:CGRectMake(0, 0, pickerView.frame.size.width, 60)];

    NSString* waveName = [((NSDictionary*)[self.myWaves objectAtIndex:row]) objectForKey:@"name"];
    NSLog(@"666666666666 titleForRow: %@", waveName);
    [label setText:waveName];
    return label;
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row   inComponent:(NSInteger)component
{
    self.waveSelected.text = [((NSDictionary*)[self.myWaves objectAtIndex:row]) objectForKey:@"name"];
    [self.waveSelected resignFirstResponder];
    
    NSLog(@",,,,,,,,,,,,,,,,,,, did select row: %@", @(row));
    APP_DELEGATE.currentWaveName = [((NSDictionary*)[self.myWaves objectAtIndex:row]) objectForKey:@"name"];
    APP_DELEGATE.currentWaveIndex = (long)row;
    //    NSLog(@"setting title: %@", APP_DELEGATE.waveName);
    
    self.navigationController.navigationBar.topItem.title = @"";//[APP_DELEGATE currentWaveName];
    [self startRefresh:self.refreshControl];
    
}


- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    NSLog(@"total images in wave: %lu", (unsigned long)self.waveImages.count);
    return [self.waveImages count];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
//    NSLog(@"@@@at index %d", indexPath.row);
    
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"ImageCell" forIndexPath:indexPath];
    
    UIImageView *waveImageView = (UIImageView *)[cell viewWithTag:100];
    
    NSString* imageName = [((NSDictionary*)[self.waveImages objectAtIndex:indexPath.row]) objectForKey:@"name"];
    NSString* waveName = [((NSDictionary*)[self.waveImages objectAtIndex:indexPath.row]) objectForKey:@"name_2"];
//    NSString* imageUrl = [NSString stringWithFormat:@"%@/img/%@/thumb_%@", EWAWSBucket, waveName, imageName];
    
    
    ((UIImageView *)[cell viewWithTag:100]).image = nil;//[UIImage imageNamed:@"echowave.png"]; // no need to show the background image

    [EWImage loadThumbImage:imageName
                    forWave:waveName
                    success:^(UIImage *image) {
                        if([collectionView.indexPathsForVisibleItems containsObject:indexPath]) {
                            ((UIImageView *)[cell viewWithTag:100]).image = image;
                            waveImageView.contentMode = UIViewContentModeScaleAspectFit;
                        }
                    } failure:^(NSError *error) {
                        NSLog(@"error: %@", error.description);
                    }];
    return cell;
}


- (void)collectionView:(UICollectionView *)imagesCollectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {

    DetailedImagePageViewController *detailedImagePageViewController = [[UIStoryboard storyboardWithName:@"Main_iPhone" bundle: nil] instantiateViewControllerWithIdentifier:@"DetailedImagePageView"];
    
    detailedImagePageViewController.initialViewIndex = indexPath.row;
    detailedImagePageViewController.waveImages = [self waveImages];
    
    [self.navigationController pushViewController:detailedImagePageViewController animated:NO];
    
}




#pragma mark -  UIPickerViewDataSource
- (NSInteger)numberOfRowsInPickerView:pickerView
{
    return [self myWaves].count;
}






@end

File: Echowaves/EchoWaveViewController.swift
//
//  EchoWaveViewController.swift
//  Echowaves
//
//  Created by D on 11/29/14.
//  Copyright (c) 2014 Echowaves. All rights reserved.
//

import Foundation

class EchoWaveViewController: UIViewController, UICollectionViewDelegate, UICollectionViewDataSource, UIPickerViewDelegate, UIPickerViewDataSource, UITextFieldDelegate {
    
    @IBOutlet var imagesCollectionView:UICollectionView!
    @IBOutlet var waveSelected:UITextField!
    @IBOutlet var emptyWaveLabel:UILabel!
    @IBOutlet var wavesPicker:UIPickerView!
    
    var waveImages = []
    var myWaves = []
    var refreshControl:UIRefreshControl!
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        NSLog("$$$$$$$$$$$$$$$$calling viewDidLoad for EchoWaveViewController")
        
        self.emptyWaveLabel.hidden = true
        
        if(self.refreshControl == nil) {
            self.refreshControl = UIRefreshControl()
            self.refreshControl.addTarget(self, action: Selector("refreshView"), forControlEvents: .ValueChanged)
            self.imagesCollectionView.addSubview(self.refreshControl)
        }    
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        self.imagesCollectionView.autoresizingMask = .FlexibleHeight
        UIViewAutoresizing.FlexibleWidth
        self.reloadWavesPicker()
    }
    
    func reloadWavesPicker() -> Void {
        EWWave.getAllMyWaves({ (waves) -> () in
            self.myWaves = waves;
            NSLog("11111111111 currentWaveName: \(APP_DELEGATE.currentWaveName)")
            
            if APP_DELEGATE.currentWaveName == "" {
                if let credential:NSURLCredential = EWWave.getStoredCredential() {
                    APP_DELEGATE.currentWaveName = credential.user!
                    APP_DELEGATE.currentWaveIndex = 0
                }
            }
            NSLog("2222222222 currentWaveName: \(APP_DELEGATE.currentWaveName)")
            NSLog("3333333333 wavesPickerSize: \(self.myWaves.count)")
            
            self.wavesPicker = UIPickerView(frame: CGRectZero)
            
            self.attachPickerToTextField(self.waveSelected, picker: self.wavesPicker)
            
            self.wavesPicker.selectRow(APP_DELEGATE.currentWaveIndex, inComponent: 0, animated: true)
            
            NSLog("setting wave index: \(APP_DELEGATE.currentWaveIndex)")
            self.navigationController?.navigationBar.topItem?.title = ""
            
            self.waveSelected.text = APP_DELEGATE.currentWaveName
            
            self.refreshView()
            
            },
            failure: { (error) -> () in
                EWWave.showErrorAlertWithMessage(error.description, fromSender: self)
        })
        
    }
    
    
    func refreshView() -> Void {
        self.startRefresh(self.refreshControl)
    }
    
    
    func attachPickerToTextField(textField:UITextField!, picker:UIPickerView!) -> Void {
        picker.delegate = self
        picker.dataSource = self
        
        textField.delegate = self
        textField.inputView = picker
    }
    
    
    func startRefresh(sender:UIRefreshControl!) -> Void {
        NSLog("starting the refresh")
        EWImage.getAllImagesForWave(APP_DELEGATE.currentWaveName,
            success: { (waveImages) -> () in
                self.waveImages = waveImages;
                NSLog("@total images \(self.waveImages.count)")
                self.imagesCollectionView.reloadData()
                self.imagesCollectionView.reloadInputViews()
                
                if self.waveImages.count == 0 {
                    NSLog("hiding false")
                    self.emptyWaveLabel.hidden = false
                } else {
                    NSLog("hiding true");
                    self.emptyWaveLabel.hidden = true
                }
                
                sender.endRefreshing()
                
            }, failure: { (error) -> () in
                NSLog("error \(error.description)")
        })
        
        
    }
    
    func numberOfComponentsInPickerView(pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return self.myWaves.count
    }
    
    func pickerView(pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusingView view: UIView!) -> UIView {
        let label = UILabel()
        label.backgroundColor = UIColor.orangeColor()
        label.textColor = UIColor.whiteColor()
        label.font = UIFont(name: "HelveticaNeue", size: CGFloat(14))
        label.textAlignment = .Center
        //WithFrame:CGRectMake(0, 0, pickerView.frame.size.width, 60)];
        
        label.text = ((self.myWaves.objectAtIndex(row) as NSDictionary).objectForKey("name") as String)
        
        return label
    }
    
    func pickerView(pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        self.waveSelected.text = ((self.myWaves.objectAtIndex(row) as NSDictionary).objectForKey("name") as String)
        self.waveSelected.resignFirstResponder()
        
        NSLog(",,,,,,,,,,,,,,,,,,, did select row: \(row)")
        APP_DELEGATE.currentWaveName = ((self.myWaves.objectAtIndex(row) as NSDictionary).objectForKey("name") as String)
        APP_DELEGATE.currentWaveIndex = row
        
        self.navigationController?.navigationBar.topItem?.title = "";//[APP_DELEGATE currentWaveName];
        self.refreshView()
    }
    
    
    
    func numberOfRowsInPickerView(pickerView:UIPickerView) -> Int {
        return self.myWaves.count
    }
    
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        NSLog("total images in wave: \(self.waveImages.count)")
        return self.waveImages.count
    }
    
    
    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        
        let cell:UICollectionViewCell  = collectionView.dequeueReusableCellWithReuseIdentifier("ImageCell", forIndexPath: indexPath) as UICollectionViewCell
        
        let waveImageView:UIImageView = cell.viewWithTag(100) as UIImageView
        
        let rowValue:NSDictionary = self.waveImages.objectAtIndex(indexPath.row) as NSDictionary
        
        let imageName = (rowValue.objectForKey("name") as String)
        let waveName = (rowValue.objectForKey("name_2") as String)
        
        
        //    ((UIImageView *)[cell viewWithTag:100]).image = nil;//[UIImage imageNamed:@"echowave.png"]; // no need to show the background image
        waveImageView.image = nil
        
        
        EWImage.loadThumbImage(
            imageName,
            waveName: waveName,
            success: { (image) -> () in
                if contains(collectionView.indexPathsForVisibleItems() as [NSIndexPath], indexPath) {
                    waveImageView.image = image
                    waveImageView.contentMode = .ScaleAspectFit
                }
            },
            failure: { (error) -> () in
                NSLog("error: \(error.description)")
            },
            progress: { (bytesRead, totalBytesRead, totalBytesExpectedToRead) -> () in
        })
        return cell;
    }
    
    
    func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        NSLog("selecting image")
        let detailedImagePageViewController:DetailedImagePageViewController! = UIStoryboard(name: "Main_iPhone", bundle: nil).instantiateViewControllerWithIdentifier("DetailedImagePageView") as DetailedImagePageViewController
        
        NSLog("index path: \(indexPath.row) and unsigned \(UInt(indexPath.row))")
        
        detailedImagePageViewController.initialViewIndex = UInt(indexPath.row)
        detailedImagePageViewController.waveImages = self.waveImages
        
        self.navigationController?.pushViewController(detailedImagePageViewController, animated: true)
    }
    
    
    
    
}